src/controlplane/scraper.go

```diff
@@ -0,0 +1,270 @@
+package controlplane
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+
+	"github.com/newrelic/infra-integrations-sdk/integration"
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/internal/discovery"
+	controlplaneClient "github.com/newrelic/nri-kubernetes/v2/src/controlplane/client"
+	"github.com/newrelic/nri-kubernetes/v2/src/controlplane/grouper"
+	"github.com/newrelic/nri-kubernetes/v2/src/scrape"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+)
+
+// Providers is a struct holding pointers to all the clients Scraper needs to get data from.
+// TODO: Extract this out of the package.
+type Providers struct {
+	K8s kubernetes.Interface
+}
+
+// Scraper takes care of getting metrics from an autodiscovered CP instances.
+type Scraper struct {
+	Providers
+	logger               log.Logger
+	config               *config.Config
+	k8sVersion           *version.Info
+	components           []component
+	informerClosers      []chan<- struct{}
+	podListerByNamespace map[string]v1.PodLister
+	inClusterConfig      *rest.Config
+}
+
+// ScraperOpt are options that can be used to configure the Scraper
+type ScraperOpt func(s *Scraper) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) ScraperOpt {
+	return func(s *Scraper) error {
+		if logger == nil {
+			return fmt.Errorf("logger canont be nil")
+		}
+		s.logger = logger
+		return nil
+	}
+}
+
+// WithRestConfig returns an OptionFunc to change the restConfig from default empty config.
+func WithRestConfig(restConfig *rest.Config) ScraperOpt {
```

Comment:
would the scraper work if WithRestConfig is not passed? would make sense to have this as a parameter?

Reply:
If I am not wrong rest config is passed to the Defalut connector that fails if it is nil, no?
```
	if inClusterConfig == nil {
		return nil, fmt.Errorf("inClusterConfig cannot be nil")
	}
```

Therefore it seems required